pkg/flow/internal/controller/loader.go

```diff
@@ -743,3 +783,30 @@ func (l *Loader) isModule() bool {
 	// Either 1 of these checks is technically sufficient but let's be extra careful.
 	return l.globals.OnExportsChange != nil && l.globals.ControllerID != ""
 }
+
+// findCustomComponentReferences returns references to import/declare nodes in a declare block.
+func (l *Loader) findCustomComponentReferences(declare *ast.BlockStmt) map[BlockNode]struct{} {
+	uniqueReferences := make(map[BlockNode]struct{})
+	l.collectCustomComponentReferences(declare.Body, uniqueReferences)
+	return uniqueReferences
+}
+
+// collectCustomComponentDependencies recursively collects references to import/declare nodes through an AST body.
+func (l *Loader) collectCustomComponentReferences(stmts ast.Body, uniqueReferences map[BlockNode]struct{}) {
+	for _, stmt := range stmts {
+		switch stmt := stmt.(type) {
+		case *ast.BlockStmt:
+			componentName := strings.Join(stmt.Name, ".")
+			switch componentName {
+			case "declare":
+				l.collectCustomComponentReferences(stmt.Body, uniqueReferences)
+			default:
+				if declareNode, ok := l.declareNodes[stmt.Name[0]]; ok {
+					uniqueReferences[declareNode] = struct{}{}
+				} else if importNode, ok := l.importConfigNodes[stmt.Name[0]]; ok {
+					uniqueReferences[importNode] = struct{}{}
+				}
+			}
+		}
+	}
```

Comment:
nit: idiomatically we try to avoid deep scope nesting like what exists here. Can you rework this a little to avoid the nesting more?

Reply:
One way to do it could be to add a function "processBlockStmt" that contains the logic 
```
        componentName := strings.Join(stmt.Name, ".")
	switch componentName {
	case "declare":
		l.collectCustomComponentReferences(stmt.Body, uniqueReferences)
	default:
		if declareNode, ok := l.declareNodes[stmt.Name[0]]; ok {
			uniqueReferences[declareNode] = struct{}{}
		} else if importNode, ok := l.importConfigNodes[stmt.Name[0]]; ok {
			uniqueReferences[importNode] = struct{}{}
		}
	}
```
The default part could also be in a function.
Else I don't see a way to put a continue or a return that could help reduce the nesting in a way that would make it more readable. Do you have a suggestion in mind?

Reply:
There's a few ways you can go about this, but if you're interested in having a switch statement, then relying on early continues helps flatten out the hierarchy: 

```go
for _, stmt := range stmts {
  blockStmt, ok := stmt.(*ast.BlockStmt)
  if !ok {
    continue 
  }

  var (
    componentName = strings.Join(blockStmt.Name, ".")

    declareNode, foundDeclare = l.declareNodes[blockStmt.Name[0]] 
    importNode, foundImport = l.importNodes[blockStmt.Name[0]]  
  )

  switch {
  case componentName == "declare":
    l.collectCustomComponentReferences(blockStmt.Body, uniqueReferences) 
  case foundDeclare:
    uniqueReferences[declareNode] = struct{}{}
  case foundImport:
    uniqueReferences[importNode] = struct{}{}
  }
}
```

If you find this to be overkill, I would be happy with just removing the outer switch statement (so that instead of the current `for -> switch -> switch -> if` it's `for -> switch -> if`).

Reply:
Looks good to me, I will go forward with this option, thanks :)